feat: Implemented project-tag based categorization for blog posts

maanasb01 opened the pull request:
This PR resolves the issue: https://github.com/tattle-made/website/issues/135

I have added project-based categorization for all blog posts. All project tags are now included with the other tags above the blog posts. If there is a need to change how project tags are rendered (e.g., creating a separate section, using a different style for project tags, or any other changes), please let me know.

**Utilizing the Reusable Component for Rendering Blogs**

The "default-blog-index-layout," "default-tag-page-layout," and "default-tag-project-page-layout" all require rendering blog posts. To streamline this, I created a separate reusable component, _AllBlogsIndexLayout_, using the _MasonryLayoutResponsive_ component (previously used in "default-blog-index-layout") to ensure consistent usage across all layouts.

This change also improves the layout of blogs rendered in the "default-tag-page-layout" (which previously used separate logic for rendering the blogs).

**Please Note:** In this commit, I have commented out the previous components' code in case it is required to revert to the old code.

dennyabrain commented:
The way this is implemented, the projects are also showing up in the tags on the index page :
![Screenshot 2024-07-28 at 20-58-58 Tattle](https://github.com/user-attachments/assets/21032634-2a20-4163-b0b6-72a6d2b90a37)
this has increased the number of tags. Would it be possible for the projects to appear in a different section titled "Projects"? They can appear right under the tag list.

maanasb01 commented:
Yeah, sure. I will implement the changes and create separate sections for projects and tags. Thanks for the feedback!

maanasb01 commented:
@dennyabrain I have committed the new changes. Please take a look.

This commit fixes the layout issue of project tags to be displayed in a different section.

I am also rendering the project tag in the specific blog page (to maintain consistency with the blog page as general tags are also being shown in each blog page), which I missed earlier. 

However, I have noticed that the project name is mentioned in the header as well, so please let me know if I should remove the project tag section.

## Created a Custom `useTags` Hook

The same code was used to create a `tagCounts` object in `default-blog-index-layout` and `default-blog-layout`. This had to be done for the project tags to show project tag counts for each tag as well. So I created a custom `useTags` hook, which provides all the tags, project tags, `tagCounts`, `projectTagCounts`, and sorted tags (same as tags but sorted by count).

I have also used this hook to display the tag counts on the tag page, which were not being shown earlier.

## Created a Reusable `TagsRenderer` Component

As a similar process had to be done to show the general tags and project tags, I have created a reusable `TagsRenderer` component to render tags and project tags. It is used in `default-blog-index-layout` and `default-blog-layout`.

## Created a Reusable `TagPage` Component

As the tag page and project tag page share the same layout, I created a reusable `TagPage` component to render these pages.

## Added a "Back to All Blogs" Link to Tag Pages

Added a "Back to All Blogs" link to the tag and project tag pages to help users navigate to the blog page (like being done in `default-blog-layout`). Also, changed the direction property from `row-responsive` to `row` to save some extra space in small devices.

## Minor: Blogs on the tag page have cover images now

Earlier, the `cover` property was not used in the Tag page to show blogs with covers. Now, I have included it in the GraphQl query. Please let me know if I need to revert this change. 

## Minor: `tags_set` in `gatsby-node`

Earlier, in `gatsby-node`, the `tags_set` was created for each node. For every node, an iteration through all the nodes was done to calculate the `tags_set` repeatedly. I shifted this operation of calculating `tags_set` to the top, before the nodes iteration starts.
